### ISSUE TITLE
 Consumption doesn't work without idle attempt before production

### ISSUE THREAD

**Issue by sibiryakov** [dpkp/kafka-python#685]
Hi guys,
the code below doesn't work. I'm probably doing something wrong, but I couldn't make it to work.
Kafka-python version is 1.1.1
and Kafka is 0.8.2.2.

```python
# -*- coding: utf-8 -*-
from kafka import KafkaConsumer, KafkaProducer, TopicPartition


def test_it():
    producer = KafkaProducer(bootstrap_servers="localhost:9092")
    consumer = KafkaConsumer(bootstrap_servers="localhost:9092", consumer_timeout_ms=100)
    consumer.assign([TopicPartition("frontera-todo", 0)])

    producer.send("frontera-todo", value="11asdhey all")
    producer.flush()

    m = next(consumer)
    
    assert m.topic == "frontera-todo" and m.value == "11asdhey all"

    producer.close()
    consumer.close()

test_it()
```

If you add this consumption attempt before ``producer.send``, then it starts to work.
```python
    try:
        m = next(consumer)
    except StopIteration:
        pass
```

Could you please explain me what's going on in this code and suggest the right way of producing/consuming from the same process.

**Comment by dpkp** [comment]
First suggestion: get the topic/partition from producer.send and verify that you produced to the correct partition, or set the partition explicitly in send(): `metadata = producer.send(...).get()`

Second, dont assume that your consumer and producers are synchronized by physical time. Set your consumer offset (logical clock) explicitly using `consumer.seek()`

**Comment by sibiryakov** [comment]
metadata shows it was produced to partition 0 (as expected).
> RecordMetadata(topic=frontera-todo, partition=0, offset=4)

`consumer.seek_to_end()` or `consumer.seek_to_beginning()` before consumption has no effect.

**Comment by sibiryakov** [comment]
Another issue: ``seek()`` is only available to assigned partitions in case of subscription, the assert is thrown.

**Comment by dpkp** [comment]
Correct - seek() requires an assigned partition. It will work on subscriptions but only after the assignment phase.

You might also look at passing a ConsumerRebalanceListener to subscribe() and use that to trigger seeks

**Comment by sibiryakov** [comment]
So the results of about 4 hours of experimenting and code reading.
I had to find a way to initialize consumer offset cache without calling ``next(consumer)``, because it could end up with message loss (or ugly buffering to avoid it).

In order to initialize the consumer offset cache from client side I found two ways:
First one is paste from iterator initialization:
```python
    def _init_consumer(self):
        if self._consumer._use_consumer_group():
            self._consumer._coordinator.ensure_coordinator_known()
            self._consumer._coordinator.ensure_active_group()

        # 0.8.2 brokers support kafka-backed offset storage via group coordinator
        elif self._consumer.config['group_id'] is not None and self._consumer.config['api_version'] >= (0, 8, 2):
            self._consumer._coordinator.ensure_coordinator_known()

        # fetch offsets for any subscribed partitions that we arent tracking yet
        if not self._consumer._subscription.has_all_fetch_positions():
            partitions = self._consumer._subscription.missing_fetch_positions()
            self._consumer._update_fetch_positions(partitions)
```

and next one is 
```
        for tp in self._partition_ids:
            self._consumer.committed(tp)
        self._consumer._update_fetch_positions(self._partition_ids)
```

Both of them aren't intuitive and requiring user to have quite in-depth understanding of consumer internal processes.

I propose to call the above mentioned code from KafkaConsumer ``assign()`` or ``subscribe()`` so consumer will be properly initialized right after assignment/subscription phase.
